pennylane/gradients/hessian_transform.py

```diff
@@ -0,0 +1,235 @@
+        _wrapper = super().default_qnode_wrapper(qnode, targs, tkwargs)
+        cjac_fn = qml.transforms.classical_jacobian(qnode, expand_fn=self.expand_fn)
+
+        def hessian_wrapper(*args, **kwargs):
+            qhess = _wrapper(*args, **kwargs)
+
+            if any(m.return_type is qml.operation.Probability for m in qnode.qtape.measurements):
+                qhess = qml.math.squeeze(qhess)
+
+            if not hybrid:
+                return qhess
+
+            kwargs.pop("shots", False)
+            cjac = cjac_fn(*args, **kwargs)
+
+            if isinstance(cjac, tuple):
+                # Classical processing of multiple arguments is present. The classical Jacobian is
+                # a tuple of Jacobians, each with respect to one QNode argument, and each of shape:
+                #   (# gate args, qnode arg shape)
+                # The quantum Hessian qhess has shape:
+                #   (qnode output shape, # gate args, # gate args)
+                # Return cjac.T @ qhess @ cjac (in the scalar case).
+                hessians = []
+
+                for c in cjac:
+                    if c is not None:
+                        # number of dimensions in qnode arg shape
+                        num_arg_dims = len(qml.math.shape(c)) - 1
+                        # hess -> (qnode output shape, # gate args, qnode arg shape)
+                        hess = qml.math.tensordot(qhess, c, [[-1], [0]])
+                        # hess -> (qnode output shape, qnode arg shape, qnode arg shape)
+                        hess = qml.math.tensordot(hess, c, [[-1 - num_arg_dims], [0]])
+                        hessians.append(hess)
+
+                return tuple(hessians)
+
+            is_square = cjac.shape == (1,) or (cjac.ndim == 2 and cjac.shape[0] == cjac.shape[1])
+
+            if is_square and qml.math.allclose(cjac, qml.numpy.eye(cjac.shape[0])):
+                # Classical Jacobian is the identity. No classical processing
+                # is present inside the QNode.
+                return qhess
+
+            # Classical processing present of either:
+            #   a) a single argument or
+            #   b) multiple arguments of the same shape
+            # The shape of the classical jacobian returned by qml.jacobian (gets invoked by
+            # qml.transforms.classical_jacobian with autograd) depends on the scenario:
+            #   a) (# gate args, qnode arg shape)
+            #   b) (reverse qnode arg shape, # gate args, # qnode args)
+            # It then needs to be contracted twice (cjac.T @ hess @ cjac for scalar-valued QNodes)
+            # with the quantum Hessian of shape:
+            #      (qnode output shape, # gate args, # gate args)
+            # The result should have shape:
+            #   a) (qnode output shape, qnode arg shape, qnode arg shape)
+            #   b) (reverse qnode arg shape A, # qnode args B, qnode output shape, qnode arg shape B, # qnode args A)
```

Comment:
This is super helpful for reading the logic below 💯 

Quick question: will any changes be needed once #2059 is merged? My understanding is no, since `qml.jacobian` is not used here, but just wanted to double check.

Reply:
Wait, `classical_jacobian` is a special case of `qml.jacobian`, right? I think this will be simplified, similar to the code in `gradient_transform`, once #2059 is merged in (where this will be done depends on which PR is merged in first, I guess?)

Reply:
Yes exactly, `qml.jacobian` is invoked by `qml.transforms.classical_jacobian`.

Reply:
Ahhh of course 🙂 Thanks for the clarification @dime10 @dwierichs!